openlibrary/plugins/openlibrary/js/goodreads_import.js

```diff
@@ -52,76 +48,108 @@ export function initGoodreadsImport() {
+                return $(`[isbn=${value['ISBN']}]`).hasClass('import-failure');
+            };
+            var fail = function (reason) {
+                if (!has_failure()) {
+                    const element = $(`[isbn=${value['ISBN']}]`);
+                    element.append(`<td class="error-imported">Error</td><td class="error-imported">${reason}</td>'`)
+                    element.removeClass('selected');
+                    element.addClass('import-failure');
+                }
+            };
+
+            if (!checked) {
+                return true;
+            }
+
+            if (shelves[shelf]) {
                 shelf_id = shelves[shelf];
             }
-            if (checked && shelf_id !== 0) {
-                prevPromise = prevPromise.then(function () { // prevPromise changes in each iteration
-                    $(`[isbn=${value['ISBN']}]`).addClass('selected');
-                    return getWork(value['ISBN']); // return a new Promise
-                }).then(function (data) {
-                    var obj = JSON.parse(data);
-                    $.ajax({
-                        url: `${obj['works'][0].key}/bookshelves.json`,
-                        type: 'POST',
-                        data: {
-                            dont_remove: true,
-                            edition_id: obj['key'],
-                            bookshelf_id: shelf_id
-                        },
-                        datatype: 'json',
-                        success: function () {
-                            if (value['My Rating'] !== '0') {
-                                $.ajax({
-                                    url: `${obj['works'][0].key}/ratings.json`,
-                                    type: 'POST',
-                                    data: {
-                                        rating: parseInt(value['My Rating']),
-                                        edition_id: obj['key'],
-                                        bookshelf_id: shelf_id
-                                    },
-                                    datatype: 'json',
-                                    success: function () {
-                                        $(`[isbn=${value['ISBN']}]`).append('<td class="success-imported">Imported</td>')
-                                        $(`[isbn=${value['ISBN']}]`).removeClass('selected');
-                                    },
-                                    fail: function () {
-                                        $(`[isbn=${value['ISBN']}]`).append('<td class="error-imported">Error</td><td class="error-imported">Failed to add Rating</td>')
-                                        $(`[isbn=${value['ISBN']}]`).removeClass('selected');
-                                    }
-                                });
+            if (shelf_id === 0) {
+                fail('Book in different Shelf');
+                func1(++count);
+                return true;
```

Comment:
```suggestion
                return false;
```
JQuery documentation recommends returning false to exit `.each()` early.